### ISSUE TITLE
Notifications  not working reliable anymor,e when app is in background

### ISSUE THREAD

**Issue by michaelof** [SimpleMobileTools/Simple-Calendar#815]
Android Marshmallow 6.0
SimpleCalendar 6.3.2

Hi Tibor,

not sure with which version, but it seems that you have changed something substantially regarding notifications, recently. Notifications worked absolutely fine for a long time. Then, not long ago, I've had to disable battery optimization to get them again. 

Now, I'm very often not getting notifications anymore if app is in background, even without battery optimization, a pity. 

For sure I have a cheap device with not much RAM (1 GB), but on this device other FOSS apps like LinPhone (VoIP client, so call notifications/ringing), K9 email client (new email notifications) etc. or closed source apps like Kakaotalk (Messenger) are working reliable on my device.

Just an idea: Not sure how SimpleCalendar works, but e.g. Mozilla Thunderbird stores for each event if any reminder has been "dismissed" or not. When starting Thunderbird, you'll get all not dismissed reminders, even if they're in the past. What about letting SimpleCalendar do the same?

**Comment by tibbi** [comment]
I dont plan storing them yet, maybe in the future. Anyway systems can block notifications in many different ways, check https://appgenix.uservoice.com/knowledgebase/articles/1827433-why-are-my-reminders-not-working if there is anything helpful.

**Comment by michaelof** [comment]
@tibbi thx for this link, MAYBE it's an "autostart" issue. I just remembered that I've had to reboot my device a couple of times recently, and I've recognized e.g. for K9 Mail, that unless started once manually, no IMAP push connections are established.

As I'm not aware of any autostart system settings, at least not for Android Marshmallow, and as web found are leading into the direction that autostart seems to be something to be requested by the app developer, is Simple Calendar auto-started?

**Comment by michaelof** [comment]
Can confirm now that this is not (only) an autostart issue. Started Simple Calendar, created an event for testing purposes: notification are triggered. Started some other (memory-itensive) apps like Firefox: Notifications are still triggered. 
But just now I've recognized, that my PC Thunderbird triggered a notification for an already existing event, starting tomorrow 2 PM, notification 1 day before. Simple Calendar does NOT trigger a notification, for the same event :-(

So IHMO this could not be affected by me with any other settings, battery optimization is off for longer now, notification for Simple Calendar are allowed. 

Question: As notifications worked without any issues for older versions of of Simple Calendar, did change anything substantial regarding notifications, recently?

**Comment by tibbi** [comment]
I dont really remember changing anything notification related lately

**Comment by ghtux** [comment]
Two different phones with lineage here (android 9). No led and no notification sound since the last two updates. The reminder itself works flawlessly.

**Comment by vocabularasa** [comment]
Yeah, I just wanted to bump this. Same issue since my Xiaomi Mi 9T Pro upgraded to MIUI Global 11 build (based on Android 10). No sounds from Event Reminders in any "audio stream" setting. All battery saving, OS optimization turned off. All app permissions, Autostart, etc on.

Simple Calendar v 6.7.2 Pro

**Comment by tibbi** [comment]
try changing the reminder audio in the app settings. If you schedule an event 1 minute into the future and keep the app foregrounded, does it work?

**Comment by vocabularasa** [comment]
In addition to attempting every available "audio stream", I also attempted several different reminder sounds, including "default", from among those listed. I also scheduled the event 1 minute head repeatedly while testing. I kept the app in the foreground, in the background, and closed altogether for debugging each "audio stream" option. No audio came through at all with any attempt.

The notifications will vibrate if I have that setting chosen, and I even tried turning vibration off thinking maybe that was overriding audio notification, but that also failed.

**Comment by vocabularasa** [comment]
I also tried setting the reminder notification and then restarting the phone, but that didn't work, either.

I also tried using the free version of Simple Calendar that's still available in Google Play. With it, I tested every available "audio stream" option and whether the app was in the foreground, background, or closed. None produced any audio reminders.

**Comment by roooots** [comment]
Hi,
I have to give this a bump again.
I bought the Pro version and generelly like it, however I have to confirm that the app has a serious issue with the reliability of reminders. For me, especially with periodically repeating events (daily, weekly), the reminders usually won't work - neither vibration, sound nor popup. Which means - reminders do work in some cases, but I did not find any clear criteria which makes most of them fail. Interestingly, the reminders won't fail if I set a test appointment several minutes or hours ahead. Another issue which affects reminders that do work is that they won't turn on the screen to show the reminder, whereas for other apps under the same circumstances this works flawlessly. I checked all necessary permissions, battery saving stuff and different notification channels - no change. I'm running the 6.9.3 on Android 9. Any suggestions?

**Comment by vassilisw** [comment]
In my Xperia Sony Android 6, notifications are not working when the event is set to be repeated every month. Hope this helps the developer.

**Comment by stoically** [comment]
Experienced the same, but realized for me it's because of Android 10's "Do not disturb" and "Battery optimization" features. Had to do two things to make it work:

1. Exempt calendars from DND
  - Go into the "App info" (either through long-touching the icon and opening "App info" or through Android Settings > Apps & notifications)
  - Touch "Notifications" in Simple Calendar' App Info
  - Make sure the "Show notifications" toggle and toggles per calendar are switched on
  - Then per calendar you want to receive notifications while DND:
    - Touch the name
    - Touch "Advanced"
    - Switch on "Override Do Not Disturb"

2. Exempt from battery optimization
  - Open Android Settings > Apps & Notifications > Advanced > Special app access > Battery optimization > Switch from "Not optimized" to "All apps"
  - Search & touch Calendar
  - Select "Don't optimize"

Note that if you want to reproduce notifications not firing with DND on, that the app must go fully into background first, which can take some time

**Comment by stoically** [comment]
Just for the sake of completeness: Turns out I was wrong, exempting from DND and battery optimization still sometimes doesn't trigger notifications. Edited my last comment accordingly.

**Comment by llevrel** [comment]
Hi,

I have found a reliable sequence to trigger the problem (I'm using v6.13.7 on Android 8.1.0):
- create an future event with reminder;
- turn off SC (Settings > Apps > SC > force off)
- launch SC
=> there will be no notification.
I tried a locally stored event or an event in one of my CalDAV calendars, same problem.

Maybe is it related to #1389 ?

**Comment by tibbi** [comment]
some tips can be found at https://dontkillmyapp.com/

**Comment by llevrel** [comment]
Thanks Tibor for answering. However in my test sequence SC is still in foreground when it should ring! It looks like SC fails/forgets to read existing events when it starts, although it does read them at system boot I suppose? (I did not test a reboot yet, this makes me nervous since my recent boot-loop problem...)

**Comment by llevrel** [comment]
> Hi,
> 
> I have found a reliable sequence to trigger the problem (I'm using v6.13.7 on Android 8.1.0):
> 
> * create a future event with reminder;
> * turn off SC (Settings > Apps > SC > force off)
> * launch SC
>   => there will be no notification.
>   I tried a locally stored event or an event in one of my CalDAV calendars, same problem.
> 
> Maybe is it related to #1389 ?

Just installed v6.19.0 from F-Droid.

The bug is still here. This is a serious bug! Tibor, can't you reproduce it following the steps above?

**Comment by tibbi** [comment]
check your device battery and notification settings, if there is nothing blocking the reminders, or killing the app in the background. You might also want to take a look at https://dontkillmyapp.com , it has some useful tips.

**Comment by llevrel** [comment]
Hi Tibor,

Notifications are not blocked, they sometimes work. I'm giving here a "steps-to-reproduce" that will consistently yield the problem. Did you test it? **SC is in the foreground**

**Comment by tibbi** [comment]
no, but you are not supposed to force stop apps anyway, that can easily break stuff

**Comment by tibbi** [comment]
thats a whole different behaviour, phone reboot is ok

**Comment by llevrel** [comment]
Testing with v6.15.2 Pro

1. In SC, create event with reminder in a CalDAV calendar account
2. Reboot phone
→ no notification by SC (Google Agenda does)

1. In SC, create event with reminder in a CalDAV calendar account
2. Sync with DAVx5
3. Reboot phone
→ no notification by SC (Google Agenda does)

1. In SC, create event with reminder in a CalDAV calendar account
2. Reboot phone
3. Launch SC and keep in foreground
→ no notification by SC (Google Agenda does)

1. In Evolution, create event with reminder in a CalDAV calendar account
2. Sync with DAVx5
3. Sync CalDAV in SC → the event shows up in the view
4. Reboot phone
→ no notification by SC (Google Agenda does)

1. In Evolution, create event with reminder in a CalDAV calendar account
2. Sync with DAVx5
3. Sync CalDAV in SC → the event shows up in the view
4. Reboot phone
5. Launch SC and keep in foreground
→ no notification by SC (Google Agenda does)
